storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1666 @@ buf_block_t *fsp_get_latched_xdes_page(
+    buf_block_t *last_descr= buf_page_get_gen(page_id_t{m_space->id, 0}, 0,
+                                              RW_S_LATCH, nullptr,
+                                              BUF_GET_POSSIBLY_FREED, &mtr,
+                                              &err);
+    if (!last_descr)
+    {
+func_exit:
+      mtr.commit();
+      return err;
+    }
+
+    for (uint32_t xdes_n= 0; xdes_n < m_space->free_limit;
+         xdes_n+= FSP_EXTENT_SIZE)
+    {
+      /* Ignore doublewrite buffer extent */
+      if (m_space == fil_system.sys_space
+          && buf_dblwr.is_inside(xdes_n))
+        continue;
+      uint32_t descr_page_no=
+        xdes_calc_descriptor_page(m_space->id, xdes_n);
+      if (descr_page_no != last_descr_page_no)
+      {
+        last_descr= buf_page_get_gen(page_id_t{m_space->id, xdes_n},
+                                     0, RW_S_LATCH, nullptr,
+                                     BUF_GET_POSSIBLY_FREED, &mtr,
+                                     &err);
+	if (!last_descr)
+          goto func_exit;
+      }
+      xdes_t *descr= XDES_ARR_OFFSET + XDES_SIZE
+        * xdes_calc_descriptor_index(0, xdes_n)
+	  + last_descr->page.frame;
+      last_descr_page_no= descr_page_no;
+      /* Ignore the extent descriptor extent */
+      if (xdes_n % srv_page_size == 0 && xdes_get_n_used(descr) == 2)
+        continue;
+      m_extent_info[xdes_n]= uint32_t(xdes_get_state(descr));
+    }
+    goto func_exit;
+  }
+
+  /** Find the earlier free extent for the given used extent
+  @param max_limit Find the extent below max limit extent
+  @return value or FIL_NULL if there is no extent */
+  uint32_t find_free_extent(uint32_t max_limit) noexcept
+  {
+    for (auto &extent_info : m_extent_info)
+    {
+      if (max_limit <= extent_info.first)
+        return FIL_NULL;
+
+      /* Ignore the already used extents */
+      if (extent_info.second >> 31)
+        continue;
+      if (extent_info.second == XDES_FREE)
+      {
+        /* Mark the extent as used one */
+        extent_info.second |= (1U << 31);
+        return extent_info.first;
+      }
+    }
+    return FIL_NULL;
+  }
+
+  /** Defragment the indexes */
+  dberr_t defragment_index(dict_index_t *index) noexcept
+  {
+    indexDefragmenter index_defrag(index);
+    return index_defrag.defragment(this);
+  }
+
+  /** Defragment the table */
+  dberr_t defragment_table(const dict_table_t *table) noexcept
+  {
+    for (dict_index_t *index= dict_table_get_first_index(table);
+         index; index= dict_table_get_next_index(index))
+    {
+      dberr_t err= fseg_validate(fil_system.sys_space, index);
+      if (err == DB_SUCCESS)
+        err= defragment_index(index);
+
+      if (err) return err;
+    }
+    return DB_SUCCESS;
+  }
+public:
+  spaceDefragmenter(fil_space_t *space) noexcept
+    : m_space(space), m_extent_size(FSP_EXTENT_SIZE) {}
+
+  fil_space_t *space() { return m_space; }
+
+  uint32_t extent_size() { return m_extent_size; }
```

Comment:
What’s the point of this member function? Why can’t we just declare these data members `const` and `public`? They are initialized by the constructor.